src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java

```diff
@@ -781,6 +786,157 @@ private void mergeStagingTable(TableId stagingTableId, TableBlob blob,
+    if (sourceEventOrdering == SourceProperties.Ordering.UN_ORDERED) {
+      /*
+       * Query will be of the form:
+       *
+       * SELECT A.* FROM
+       *   (SELECT * FROM [staging table] WHERE _batch_id = 1234567890 AND _sequence_num > $LATEST_APPLIED) as A
+       *   LEFT OUTER JOIN
+       *   (SELECT * FROM [staging table] WHERE _batch_id = 1234567890 AND _sequence_num > $LATEST_APPLIED) as B
+       *   ON A._row_id = B._row_id AND A._sequence_num < B._sequence_num
+       *   WHERE B._row_id IS NULL
+       *
+       * Here we only construct the join condition and where clause as that is the only difference
+       * based on ordering of events.
+       */
+
+      joinCondition =  "A._row_id = B._row_id ";
+      whereClause = " B._row_id IS NULL ";
+    } else {
+      /*
+       * Query will be of the form:
+       *
+       * SELECT A.* FROM
+       *   (SELECT * FROM [staging table] WHERE _batch_id = 1234567890 AND _sequence_num > $LATEST_APPLIED) as A
+       *   LEFT OUTER JOIN
+       *   (SELECT * FROM [staging table] WHERE _batch_id = 1234567890 AND _sequence_num > $LATEST_APPLIED) as B
+       *   ON A.id = B._before_id AND A._sequence_num < B._sequence_num
+       *   WHERE B._before_id IS NULL
+       *
+       * Here we only construct the join condition and where clause as that is the only difference
+       * based on ordering of events.
+       */
+      joinCondition = primaryKeys.stream()
+        .map(name -> String.format("A.%s = B._before_%s", name, name))
+        .collect(Collectors.joining(" AND "));
+
+      whereClause = primaryKeys.stream()
+        .map(name -> String.format("B._before_%s IS NULL", name))
+        .collect(Collectors.joining(" AND "));
+    }
+
+    return "SELECT A.* FROM\n" +
+      "(SELECT * FROM " + stagingTable.getDataset() + "." + stagingTable.getTable() +
+      " WHERE _batch_id = " + batchId +
+      " AND _sequence_num > " + latestSequenceNumInTargetTable + ") as A\n" +
+      "LEFT OUTER JOIN\n" +
+      "(SELECT * FROM " + stagingTable.getDataset() + "." + stagingTable.getTable() +
+      " WHERE _batch_id = " + batchId +
+      " AND _sequence_num > " + latestSequenceNumInTargetTable + ") as B\n" +
+      "ON " + joinCondition + " AND A._sequence_num < B._sequence_num\n" +
+      "WHERE " + whereClause;
+  }
+
+  static String createMergeQuery(TableId targetTableId, List<String> primaryKeys, Schema targetSchema,
+                                 String diffQuery, SourceProperties.Ordering sourceEventOrdering) {
+    String mergeCondition;
+    String deleteOperation;
+
+    if (sourceEventOrdering == SourceProperties.Ordering.UN_ORDERED) {
+      /*
+       * Merge query will be of the following form:
+       *
+       * MERGE [target table] as T
+       *  USING ($DIFF_QUERY) as D
+       * ON T._row_id = D._row_id
+       * WHEN MATCHED AND D.op = “DELETE”
+       *    UPDATE _is_deleted = true
+       * WHEN MATCHED AND D.op IN (“INSERT”, “UPDATE”)
```

Comment:
Do we need "INSERT" here ?  In which case an INSERT op will match an existing row ?

Reply:
Its possible in the scenario where in the same batch if the record with primary key x is deleted and new record with the same primary key is inserted. New record will have _op as INSERT.

Reply:
in this case , you won't be able to MATCH. for ORDERED case, INSERT event won't have value of "_before" columns. for UNORDERED case, row_id will be different between the newly inserted row and old deleted row.